akka-projection-grpc/src/main/scala/akka/projection/grpc/replication/scaladsl/ReplicatedEventSourcingOverGrpc.scala

```diff
@@ -0,0 +1,133 @@
+
+    // sharding for forwarding events
+    val sharding = ClusterSharding(system)
+    val entity = Entity[Command](settings.entityTypeKey) { entityContext =>
+      val replicationId =
+        ReplicationId(settings.entityTypeKey.name, entityContext.entityId, settings.selfReplicaId)
+      ReplicatedEventSourcing.externalReplication(replicationId, allReplicaIds)(replicatedBehaviorFactory)
+    }
+    sharding.init(entity)
+
+    // sharded daemon process for consuming event stream from the other dc:s
+    settings.otherReplicas.foreach { replica =>
+      startConsumer(
+        settings.entityTypeKey.name,
+        replica,
+        settings.protobufDescriptors,
+        sharding.entityRefFor(entity.typeKey, _))
+    }
+
+    // FIXME user still has to start the producer, could we opt-out do that as well?
+    new ReplicatedEventSourcingOverGrpc[Command](eventProducerRoute, entity.typeKey)
+  }
+
+  private def startConsumer[C](
+      entityTypeKeyName: String,
+      replica: Replica,
+      protobufDescriptors: Seq[Descriptors.FileDescriptor],
+      entityRefFactory: String => EntityRef[C])(implicit system: ActorSystem[_]): Unit = {
+    implicit val timeout: Timeout = 5.seconds // FIXME from config
+
+    val projectionName = s"${entityTypeKeyName}_${replica.replicaId.id}"
+    val sliceRanges = Persistence(system).sliceRanges(replica.numberOfConsumers)
+
+    val eventsBySlicesQuery =
+      GrpcReadJournal(replica.grpcQuerySettings, replica.grpcClientSettings, protobufDescriptors)
+
+    ShardedDaemonProcess(system).init(projectionName, replica.numberOfConsumers, {
+      idx =>
+        val sliceRange = sliceRanges(idx)
+        val projectionKey =
+          s"${eventsBySlicesQuery.streamId}-${sliceRange.min}-${sliceRange.max}"
+        val projectionId = ProjectionId.of(projectionName, projectionKey)
+
+        val replicationFlow = FlowWithContext[EventEnvelope[_], ProjectionContext].mapAsync(1) { envelope =>
+          val entityRef = entityRefFactory(envelope.persistenceId).asInstanceOf[EntityRef[PublishedEventImpl]]
+          entityRef.ask[Done](
+            replyTo =>
+              PublishedEventImpl(
+                PersistenceId.ofUniqueId(envelope.persistenceId),
+                envelope.sequenceNr,
+                envelope.event,
+                envelope.timestamp,
+                // FIXME actually provide the right metadata here
+                envelope.eventMetadata.map(_.asInstanceOf[ReplicatedPublishedEventMetaData]),
+                Some(replyTo)))
+        }
+
+        val sourceProvider = EventSourcedProvider.eventsBySlices[AnyRef](
+          system,
+          eventsBySlicesQuery,
+          eventsBySlicesQuery.streamId,
+          sliceRange.min,
+          sliceRange.max)
+
+        ProjectionBehavior(
+          // FIXME support more/arbitrary projection impls?
+          R2dbcProjection.atLeastOnceFlow(projectionId, None, sourceProvider, replicationFlow))
```

Comment:
Do we need the Flow, or can we use `atLeastOnceAsync`? A good reason for using flow could be that we can have parallelism > 1 for the mapAsync.

Reply:
to avoid the r2dbc dependency we would have to take a factory function as parameter to `def grpcReplication` ?

Reply:
maybe it's fine with the optional dependency, but then we have a true cycle between the projects. previously it was only tests/samples

Reply:
I think it would anyway be good to move the r2dbc projection over here.

Reply:
WRT the first question, yes, we definitely want higher parallelism since we send those events out over sharding to the entities.

Reply:
Working parallelism added in ef8f50d3621bb65150df1c7bbe6584e1431e3cf5 and r2dbc dependency pulled out/made pluggable in f6c8b3c3d0c8f5d40f752477bec7370301c06936